pkg/pillar/zedcloud/authen.go

```diff
@@ -0,0 +1,508 @@
+		log.Infof("signAuthData: tpm sigres (len %d): %x\n", len(sigres), sigres)
+	case *ecdsa.PrivateKey:
+		r, s, err := ecdsa.Sign(rand.Reader, key, hash)
+		if err != nil {
+			log.Infof("signAuthData: ecdsa sign error %v\n", err)
+			return nil, err
+			//log.Fatal("ecdsa.Sign: ", err)
+		}
+		log.Debugf("r.bytes %d s.bytes %d\n", len(r.Bytes()),
+			len(s.Bytes()))
+		sigres = r.Bytes()
+		sigres = append(sigres, s.Bytes()...)
+		log.Infof("signAuthData: ecdas sigres (len %d): %x\n", len(sigres), sigres)
+	}
+	return sigres, nil
+}
+
+// ComputeSha - Compute sha256 on data
+func ComputeSha(data []byte) []byte {
+	h := sha256.New()
+	h.Write(data)
+	hash := h.Sum(nil)
+	return hash
+}
+
+func checkMimeProtoType(r *http.Response) bool {
+	if r == nil {
+		return false
+	}
+	var ctTypeProtoStr = "application/x-proto-binary"
+	ct := r.Header.Get("Content-Type")
+	if ct == "" {
+		return false
+	}
+	mimeType, _, err := mime.ParseMediaType(ct)
+	if err != nil {
+		return false
+	}
+
+	return mimeType == ctTypeProtoStr
+}
+
+// VerifyCloudCertChain - verify certificate chain from controller
+func VerifyCloudCertChain(ctx ZedCloudContext, serverName string, content []byte) error {
+	sm := &zcert.ZControllerCert{}
+	err := proto.Unmarshal(content, sm)
+	if err != nil {
+		log.Infof("SaveCertsToFile: Unmarshal err %v\n", err)
+		return err
+	}
+
+	var block *pem.Block
+	var leafcert *x509.Certificate
+	var certByte []byte
+	var keyCnt int
+	interm := x509.NewCertPool()
+	for _, cert := range sm.GetCerts() {
+		keyCnt++
+		if cert.Type == zcert.ZCertType_CERT_TYPE_CONTROLLER_INTERMEDIATE {
+			ok := interm.AppendCertsFromPEM(cert.GetCert())
+			if !ok {
+				err := fmt.Errorf("VerifyCloudCertChain: fail to append intermediate certs")
+				return err
+			}
+			log.Infof("VerifyCloudCertChain: get intermediate cert %d, len %d\n", keyCnt, len(cert.GetCert()))
+		} else if cert.Type == zcert.ZCertType_CERT_TYPE_CONTROLLER_SIGNING {
+			if len(certByte) > 0 {
+				err := fmt.Errorf("VerifyCloudCertChain: received more than one leaf cert")
+				return err
+			}
+			certByte = cert.GetCert()
+			block, _ = pem.Decode(certByte)
+			if block == nil {
+				err := fmt.Errorf("VerifyCloudCertChain: can not get cert block")
+				return err
+			}
+			leafcert, err = x509.ParseCertificate(block.Bytes)
+			if err != nil {
+				log.Infof("VerifyCloudCertChain: fail to parse certificate %v\n", err)
+				return err
+			}
+			log.Infof("VerifyCloudCertChain: get signing cert %d, len %d\n", keyCnt, len(cert.GetCert()))
+		} else {
+			err := fmt.Errorf("VerifyCloudCertChain: received cert with unknown type %d", cert.Type)
+			return err
+		}
+	}
+
+	log.Infof("VerifyCloudCertChain: key count %d\n", keyCnt)
+	if leafcert == nil {
+		err := fmt.Errorf("VerifyCloudCertChain: fail to acquire leaf cert")
+		return err
+	}
+
+	// Get the rootCAs from the zedcloudCtx
+	var roots *x509.CertPool
+	if ctx.TlsConfig != nil {
+		roots = ctx.TlsConfig.RootCAs
```

Comment:
Incorrect. You should only take the root(s) from /config/root-certificate.pem
Not all of the TLS root CAs.

Reply:
i think currently, the gamma and alpha just send the TLS certs, yes for alpha it's the /config/root-certificate.pem, but for gamma, it's still their TLS certs. Thus at least for now, this ctx.TlsConfig.RootCAs has both. If later gamma those hosts moving to 'untrusted' private certs for this, we can change to that root-certificate.pem only. I can make a comment on this here.

Reply:
As we discussed let's address this in the next sprint and get the V2 signing key to come from the /config/root-certificate.pem root.